Convert hui-vertical-stack-card to TypeScript/LitElement

iantrich opened the pull request:
Having trouble passing hass object to created elements. Things like the markdown card work find but those that have entities in hass are acting like they don't exist.

iantrich committed:
Convert hui-vertical-stack-card to TypeScript/LitElement

Having trouble passing hass object to created elements. Things like the markdown card work find but those that have entities in hass are acting like they don't exist.

iantrich committed:
Convert hui-vertical-stack-card to TypeScript/LitElement

Having trouble passing hass object to created elements. Things like the markdown card work find but those that have entities in hass are acting like they don't exist.

thomasloven commented on src/panels/lovelace/cards/hui-vertical-stack-card.ts:
```diff
@@ -0,0 +1,102 @@
+
+    this.config = config;
+
+    if (this.shadowRoot) {
```
This check isn't needed anymore. `_buildConfig()` can be moved into `setConfig`.

balloob commented on src/panels/lovelace/cards/hui-vertical-stack-card.ts:
```diff
@@ -0,0 +1,102 @@
+  static get properties() {
+    return {
+      hass: {
+        observer: "_hassChanged",
```
This doesn't exist in lit.

balloob replied:
You need to remove `hass` as a property that causes lit to re-render. Make your own updater

```
set hass(value) {
    this._hass = hass;
    this.shadowRoot.querySelectorAll('#root > *').forEach(el => { el.hass = hass; });
}
```

iantrich replied:
```
set hass(hass: HomeAssistant) {
    this.hass = hass;
    for (const el of this.shadowRoot.querySelectorAll('#root > *')) {
      el.hass = hass;
    };
  }
```

This appears to put me an infinite loop where updating `hass` of the element updates `hass` in general

zsarnett replied:
this.hass = hass; is your infinite loop

zsarnett replied:
You will need to remove it as a property

iantrich replied:
oh boy...I need more coffee for this job

balloob commented on src/panels/lovelace/cards/hui-vertical-stack-card.ts:
```diff
@@ -0,0 +1,102 @@
+    `;
+  }
+
+  private _buildConfig() {
```
Move this into the render method.

balloob commented on src/panels/lovelace/cards/hui-vertical-stack-card.ts:
```diff
@@ -0,0 +1,102 @@
+  }
+
+  private _buildConfig() {
+    this.elements = [];
```
We don't have to use `this.elements` anymore.

balloob commented on src/panels/lovelace/cards/hui-vertical-stack-card.ts:
```diff
@@ -0,0 +1,102 @@
+    return html`
+      ${this.renderStyle()}
+      <div id="root">
+        ${this.elements!.map(element => element)}
```
if you map with a function that returns itself, you don't need a map 😉

thomasloven committed:
Add option to pick number of columns for glance card (#1832)

balloob committed:
Always define localize func (#1830)

iantrich committed:
Address review comments

Need help with updating child elements when hass updates

iantrich committed:
Address review comments

Need help with updating child elements when hass updates

balloob committed:
Fix chart tooltip

zsarnett committed:
Merge pull request #1834 from home-assistant/fix-chart-tooltip

Fix chart tooltip

iantrich committed:
Properly update hass for elements

I don't like the variable name `mHass` but Linter didn't like `_hass`

iantrich committed:
Properly update hass for elements

I don't like the variable name `mHass` but Linter didn't like `_hass`

balloob committed:
Fix second visit to history panel (#1835)

balloob committed:
Reinstate first call true

balloob commented on src/panels/lovelace/cards/hui-vertical-stack-card.ts:
```diff
@@ -0,0 +1,87 @@
+  }
+
+  set hass(hass: HomeAssistant) {
+    this.mHass = hass;
```
`this._hass`

iantrich replied:
TS did not like _hass as a variable name

zsarnett replied:
It allows it with Travis I believe. Not sure if thats a good or bad thing

zsarnett replied:
Disregard this is false. Entities Conversion failed because of this

balloob replied:
let's disable that rule in TS Lint

iantrich replied:
`"variable-name": [true, "ban-keywords", "check-format", "allow-leading-underscore"]`

Does this look right?

balloob replied:
if the error stops happening: yeah 😁

balloob commented on src/panels/lovelace/cards/hui-vertical-stack-card.ts:
```diff
@@ -0,0 +1,87 @@
+    return html`
+      ${this.renderStyle()}
+      <div id="root">
+        ${this.config.cards.map(card => createCardElement(card, this.mHass))}
```
there is no second parameter?

balloob replied:
You'll need to do this above the return and then loop through it and set `el.hass = this._hass`

iantrich replied:
The linter doesn't like that I'm trying to set `hass` on something of type `Element`. Is there a defined type we have that I should be casting it `as`? I was thinking `LovelaceCard` but doesn't look like that has `hass` in it.

iantrich replied:
Also started seeing lint errors like this:
`Property 'config' has no initializer and is not definitely assigned in the constructor.`
I ran into something similar on another project with TS and ended up using classes with default initialization constructors.

balloob replied:
we want to convert createCardElement to typescript or cast it to a card:

```ts
const card = createCardElement(card) as LovelaceCard
```

balloob replied:
Make it `config?`. We don't initialize it in the constructor so should mark it as that it might not be available.

balloob commented on src/panels/lovelace/common/create-card-element.js:
```diff
@@ -69,7 +69,7 @@ function _createErrorElement(error, config) {
 }
 
-export default function createCardElement(config) {
+export default function createCardElement(config, hass = undefined) {
```
I think that this is not a good addition. It's conflating concerns. Config is standalone of `hass`.

tomraithel committed:
Improve size of device cards and add max-width for entities (#1838)

fixes #1751

iantrich committed:
Generate icons from `.ts. files (#1842)

iantrich committed:
`connection` is being populated for `HomeAssistant` type, not `conn` (#1841)

thomasloven committed:
Add HomeAssistant object hass to LovelaceCard interface (#1843)

thomasloven committed:
Introduce typescript to hui-conditional-card (#1831)

balloob committed:
TS history data (#1839)

* Convert history data to TS

* Lint

* Extract cached history

* Move around

balloob committed:
Update translations

balloob committed:
Update version to 20181023.0

cgarwood committed:
Reduces device registry card height to only show 4 entities before scrolling (#1844)

iantrich committed:
Addressed review comments

Getting the following lint errors:

`Property 'config' has no initializer and is not definitely assigned in the constructor.`
`Property '_hass' has no initializer and is not definitely assigned in the constructor.`

iantrich committed:
Remove passing of `hass` to `createCardElement`

balloob commented on src/panels/lovelace/cards/hui-vertical-stack-card.ts:
```diff
@@ -0,0 +1,95 @@
+
+  public getCardSize() {
+    let totalSize = 0;
+    for (const element of this.shadowRoot!.querySelectorAll("#root > *")) {
```
this won't work as we haven't rendered yet here. Not sure if Lit has a synchronous render function that we can use.

balloob commented on src/panels/lovelace/common/create-card-element.js:
```diff
@@ -1,108 +1,108 @@
-
-  return _createElement(`hui-${config.type}-card`, config);
-}
+import { fireEvent } from "../../../common/dom/fire_event.js";
```
What happened to this file?

iantrich replied:
Was supposed to just revert my change to it. I'll double check the contents or is there a better way to revert to before my commit to this file to be sure it hasn't inadvertently changed?

balloob commented on src/panels/lovelace/types.ts:
```diff
@@ -6,3 +8,7 @@ export interface LovelaceCard {
   setConfig(config: LovelaceConfig): void;
 }
+
+export interface LovelaceCardElement extends HTMLElement {
```
`LovelaceCardElement` is a `LovelaceCard`. If you rebase on `master`, this change has already been applied https://github.com/home-assistant/home-assistant-polymer/blob/c05b77961ebbcaae385bcb53689f2992114cc0cd/src/panels/lovelace/types.ts#L7-L11

iantrich commented:
I think once I rebase to pickup recent changes, I can easily remove my changes to the merge conflicts.

iantrich committed:
Merge branch 'typescript-vertical' of https://github.com/iantrich/home-assistant-polymer into typescript-vertical

iantrich commented:
I've botched this rebase...what is the suggested way to remedy this?

iantrich commented:
Sorry, I failed to get this rebase to work properly. I'm closing this PR and going to start fresh. I'll try to be more careful in the future.
